src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Fixer.cs

```diff
@@ -60,16 +65,53 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
 
             SyntaxGenerator generator = SyntaxGenerator.GetGenerator(doc);
 
-            // The desired case changing enum value should've already been chosen in the analyzer
-            if (!context.Diagnostics[0].Properties.TryGetValue(RCISCAnalyzer.CaseChangingApproachName, out string? caseChangingApproachValue) || caseChangingApproachValue == null)
+            ImmutableDictionary<string, string?> dict = context.Diagnostics[0].Properties;
+
+            // The dictionary should contain the keys for both left and right offending methods,
+            // and at least one of them should not be null.
+            if (!dict.TryGetValue(RCISCAnalyzer.LeftOffendingMethodName, out string? leftOffendingMethod) ||
+                !dict.TryGetValue(RCISCAnalyzer.RightOffendingMethodName, out string? rightOffendingMethod) ||
+                (leftOffendingMethod == null && rightOffendingMethod == null))
+            {
+                return;
+            }
+
+            bool leftIsToLowerOrToUpper = leftOffendingMethod is RCISCAnalyzer.StringToLowerMethodName or RCISCAnalyzer.StringToUpperMethodName;
+            bool rightIsToLowerOrToUpper = rightOffendingMethod is RCISCAnalyzer.StringToLowerMethodName or RCISCAnalyzer.StringToUpperMethodName;
+
+            bool leftIsToLowerInvariantOrToUpperInvariant = leftOffendingMethod is RCISCAnalyzer.StringToLowerInvariantMethodName or RCISCAnalyzer.StringToUpperInvariantMethodName;
+            bool rightIsToLowerInvariantOrToUpperInvariant = rightOffendingMethod is RCISCAnalyzer.StringToLowerInvariantMethodName or RCISCAnalyzer.StringToUpperInvariantMethodName;
+
+            // If the cultures of the two strings are incompatible, do not offer a fix
+            if ((leftIsToLowerOrToUpper && rightIsToLowerInvariantOrToUpperInvariant) ||
+                (rightIsToLowerOrToUpper && leftIsToLowerInvariantOrToUpperInvariant))
             {
                 return;
             }
 
+            string? caseChangingApproachValue = null;
+            if (leftIsToLowerOrToUpper || rightIsToLowerOrToUpper)
+            {
+                caseChangingApproachValue = RCISCAnalyzer.StringComparisonCurrentCultureIgnoreCaseName;
+            }
+            else if (leftIsToLowerInvariantOrToUpperInvariant || rightIsToLowerInvariantOrToUpperInvariant)
+            {
+                caseChangingApproachValue = RCISCAnalyzer.StringComparisonInvariantCultureIgnoreCaseName;
+            }
+
+            Debug.Assert(caseChangingApproachValue != null, "Unexpected offending methods");
+
             if (operation is IInvocationOperation invocation)
             {
+                if (invocation.TargetMethod.Name is RCISCAnalyzer.StringCompareToMethodName)
+                {
+                    // Never offer a fix for CompareTo
+                    return;
+                }
+
```

Comment:
would be better to check this earlier and shortcut returning?

Reply:
If I do this earlier, I would have to cast it anyway to an `IInvocationOperation` so I can check the target method name, but there's also the possibility that it is not an invocation, but an `IBinaryOperation`, which is the `else if` case here.

I think it wouldn't help much to do it earlier. The analyzer already confirmed this is a valid case, so the fixer shouldn't need to worry much about getting the wrong one.